Add partial form to select nodes for circuit proposal

eloaverona opened the pull request:
This PR adds a partial form to select nodes for a circuit proposal, and a form to add a new node to the splinter registry.

#### To test: 
1. Run:
```
docker-compose -f docker/docker-compose.yaml up --build
```
2. To see the alpha node UI go to: `localhost:3030`

3. Click on the `circuits` option in the left side menu, it will forward you to  `localhost:3030/circuits`:
<img width="1005" alt="Screen Shot 2020-06-11 at 8 15 48 AM" src="https://user-images.githubusercontent.com/14094978/84389783-333c1b80-abbc-11ea-95c6-a6bbeed42f82.png">


4. Click "Propose Circuit" button. You will be forward to the Propose Circuit Form. Test that you can select a node from the available nodes and remove a node from the selected nodes. 
<img width="1005" alt="Screen Shot 2020-06-11 at 8 15 17 AM" src="https://user-images.githubusercontent.com/14094978/84389803-39ca9300-abbc-11ea-9949-16253540c5c7.png">

5. The form is also partially responsive, in a mobile view:
<img width="825" alt="Screen Shot 2020-06-11 at 8 16 51 AM" src="https://user-images.githubusercontent.com/14094978/84389873-56ff6180-abbc-11ea-8265-1bee31671b2a.png">



P.S The other steps in the form are not implemented in this PR.

Commit:
Add proxy to circuits sapling package.json

The proxy is when running dev servers locally

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

Commit:
Route to circuit proposal form on button click

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

RyanLassigBanks reviewed (dismissed):
Tested locally

dplumb94 commented on saplings/circuits/src/components/forms/MultiStepForm.js:
```diff
@@ -0,0 +1,156 @@
+/**
```
Is this the same multi step form that is in any of the other saplings? Maybe a good opportunity to pull it out and put it in a library.

eloaverona replied:
Yes, I had to modify it a little, but it is very similar. I think it should be part of a different story to make this multi step form a library

dplumb94 commented on saplings/circuits/src/api/splinter.js:
```diff
@@ -38,3 +38,20 @@ export const getNodeRegistry = async () => {
   throw Error(result.data);
 };
+
+export const postNodeRegistry = async node => {
```
This should be called `createNode` or `addNode` I think.

eloaverona replied:
I removed the anything related to add a new node to the registry from this PR. This PR was getting too big, so there changes will be in a separate PR

dplumb94 commented on saplings/circuits/src/components/forms/ProposeCircuitForm.js:
```diff
@@ -14,18 +14,219 @@
+
+import nodeIcon from '../../images/node_icon.svg';
+
+import { Chip, Chips } from '../Chips';
```
this could get grouped with the other imports

dplumb94 commented on saplings/circuits/src/components/forms/MultiStepForm.js:
```diff
@@ -0,0 +1,156 @@
+            </button>
+          )}
+          {step < children.length && (
+            <button
```
The text on the disabled state for buttons is very difficult to read. I would recommend changing the background color too, or just making the button gray. You might also want to run the site through an accessibility checker (there is a SiteImprove accessibility checker browser extension that works well.)

eloaverona replied:
updated

dplumb94 commented on saplings/circuits/src/data/nodeRegistry.js:
```diff
@@ -0,0 +1,40 @@
+
+function NodeRegistryResponse(reponseData) {
+  this.data = reponseData.data.map(node => {
+    return new Node(node);
```
I'm curious what would happen here if the response changed in the backend and did not fit the model of this prototype.

eloaverona replied:
yeah, I probably have to add some kind of error catching here, I'll do that in a separate PR as the the circuits prototype that is already in master will have the same issue, so I'll fix them together

dplumb94 commented on saplings/circuits/src/components/forms/ProposeCircuitForm.scss:
```diff
@@ -0,0 +1,144 @@
+
+        &:hover {
+          border-width: 2px;
+          font-weight: bold;
```
I don't think the font weight needs to be bold here. Also, I think it would be a good idea to pull out the buttons into their own component that can be configured as needed, instead of creating and re-styling each button.

dplumb94 commented on saplings/circuits/src/components/Chips.js:
```diff
@@ -0,0 +1,69 @@
+
+import './Chips.scss';
+
+export function Chips({ children }) {
```
I think it would be better to style the chip container separately from the actual chip components. The styles for the "chips" class don't really have anything to do with chips specifically. There also might be unnecessary complexities if you are using the chip outside of `Chips` since the chip styles are nested.

Alternatively, If you want to be able to use `Chips`, I think the `Chips` component should take a list of data to render as chips as a prop, and then not expose the `Chip` component.

eloaverona replied:
updated

dplumb94 commented on saplings/circuits/src/components/forms/ProposeCircuitForm.js:
```diff
@@ -0,0 +1,263 @@
+            </div>
+            <ul>
+              {nodesState.filteredNodes.nodes.map(node => (
+                <li className="node-item">
```
I am finding this table to be a bit confusing. I'd recommend making this a table with headers (so the user knows what the fields are) or use some sort of card component.

eloaverona replied:
updated

dplumb94 commented on saplings/circuits/src/components/forms/ProposeCircuitForm.scss:
```diff
@@ -0,0 +1,144 @@
+      border-bottom: 1px solid var(--color-light-grey);
+      padding: 1rem 0 0.5rem 2rem;
+      background-image: url('../../images/search_icon.svg');
+      background-position: 5px 15px;
```
avoid using px

eloaverona replied:
updated

dplumb94 commented on saplings/circuits/src/components/forms/ProposeCircuitForm.js:
```diff
@@ -0,0 +1,263 @@
+                New node
+              </button>
+            </div>
+            <ul>
```
I don't know if I like that the nodes disappear from the table when selected. Maybe just add a darker background or an icon?

eloaverona replied:
updated

dplumb94 reviewed (changes requested):
Throughout the styles, there are a lot of PX used. In general, it is a best practice to use rems wherever possible to honor the user's font settings. With borders it is ok, but I'd recommend switching to rems/ems everywhere else if possible.

Commit:
Add multi-step form component

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

Commit:
Add request to get nodes from splinter registry 

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

Commit:
Add basic layout for propose circuit form

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

Commit:
Add node registry state

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

Commit:
Add node chips component

This component will represent selected
nodes in the "Add nodes" step of the circuit proposal form

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

Commit:
Add node card component

This component will display
nodes from the registry in the "Add nodes"
step of the circuit proposal form

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

Commit:
Add form elements to select nodes in the circuit proposal form

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

peterschwarz commented on saplings/circuits/src/components/forms/ProposeCircuitForm.js:
```diff
@@ -14,18 +14,285 @@
+
+const filterNodes = (state, input) => {
+  const filteredNodes = state.availableNodes.filter(node => {
+    if (node.identity.toLowerCase().indexOf(input) > -1) {
```
input should probably be made lower case here, versus in the dom event handler.  Otherwise, this function has to assume that the input is lower case to function properly.

eloaverona replied:
updated

peterschwarz commented on saplings/circuits/src/components/forms/ProposeCircuitForm.js:
```diff
@@ -14,18 +14,285 @@
+      }
+      return true;
+    }
+    if (node.displayName.toLowerCase().indexOf(input) > -1) {
```
another instance of input here.  Probably should just `let lowerInput = input.toLowerCase()` at the beginning of the function and use that.

eloaverona replied:
ah! yes, my bad. I updated it!

peterschwarz reviewed (dismissed):
Tested

dplumb94 commented on saplings/circuits/src/components/Chips.js:
```diff
@@ -0,0 +1,81 @@
+
+Chips.propTypes = {
+  nodes: PropTypes.arrayOf(Node).isRequired,
+  localNodeID: PropTypes.instanceOf(String),
```
What was the rationale for not using `PropTypes.string`?

eloaverona replied:
None, I changed it to PropTypes.string.

dplumb94 commented on saplings/circuits/src/components/Chips.js:
```diff
@@ -0,0 +1,81 @@
+    return '';
+  }
+  return (
+    <div className="chip-group">
```
I'm not quite sure how to interpret this structure, is the chip-group surrounding an individual chip?

eloaverona replied:
my bad, in refactoring things chip-group ended up becoming obsolete and I forgot to remote it.

dplumb94 commented on saplings/circuits/src/components/Chips.js:
```diff
@@ -0,0 +1,81 @@
+}
+
+function Chip({ node, isLocal, removeFn, deleteable }) {
+  if (!node) {
```
You could just make node optional and set `''` to the default prop here I believe.

eloaverona replied:
this check is not relevant anymore, I removed it

dplumb94 commented on saplings/circuits/src/components/NodeCard.js:
```diff
@@ -0,0 +1,78 @@
+          </div>
+        </div>
+        <div className="node-labels">
+          <div className={isLocal ? 'node-local' : 'not-visible'}>Local</div>
```
This div should just be conditionally rendered instead of hidden in the css:

`{isLocal && <div className="node-local">Local</div>}`

eloaverona replied:
updated

dplumb94 commented on saplings/circuits/src/components/NodeCard.scss:
```diff
@@ -0,0 +1,125 @@
+  padding: 1rem;
+  width: 100%;
+
+  .not-visible {
```
This style might not be necessary (see prev. comment) but usually you want to use `display: none` rather than `visibility: hidden`. With `visibility: hidden` the element still takes up space but is just invisible which can cause issues (there was a bug in Gameroom due to that). So I'd recommend only using the visibility property if you're sure thats what you want.

eloaverona replied:
In this case, visibility: hidden makes sense. Otherwise the checkmark to select a node will not be in the same place for local and not local nodes.

dplumb94 commented on saplings/circuits/src/components/NodeCard.scss:
```diff
@@ -0,0 +1,125 @@
+        height: 1.5rem;
+        min-height: 1.5rem;
+        border: 1px solid #555;
+        color: #555;
```
does this color correspond to any of the variables? it is good to keep the sapling colors as vars to correspond if theyre pulled into another app.

eloaverona replied:
yes, there is a variable. but in any case, those styles were obsolete so I just removed them

dplumb94 commented on saplings/circuits/src/components/forms/ProposeCircuitForm.js:
```diff
@@ -14,18 +14,286 @@
+
+      let error = '';
+      if (selectedNodes.length < 2) {
+        error = errorMessage;
```
This error message should probably be specific to the actual error

eloaverona replied:
updated

dplumb94 reviewed (changes requested):
Found a build warning while building:

```  
Line 76:6:  React Hook useEffect has missing dependencies: 'sortCircuitsBy', 
'sortedBy.ascendingOrder', and 'sortedBy.field'. Either include them or remove the dependency array  
react-hooks/exhaustive-deps
```

Commit:
Add dependencies to useEffect to sort circuits

Missing dependencies were causing a build error

Signed-off-by: Eloá Franca Verona <eloafran@bitwise.io>

eloaverona commented:
I added a commit to fix this

> Found a build warning while building:
Line 76:6:  React Hook useEffect has missing dependencies: 'sortCircuitsBy', 
'sortedBy.ascendingOrder', and 'sortedBy.field'. Either include them or remove the dependency array  
react-hooks/exhaustive-deps

dplumb94 reviewed (approved):
tested, lgtm
